File: catalogue-application/src/main/java/io/quartic/catalogue/datastore/GoogleDatastoreBackend.java
package io.quartic.catalogue.datastore;

import com.codahale.metrics.health.HealthCheck;
import com.google.api.client.util.Maps;
import com.google.cloud.datastore.Datastore;
import com.google.cloud.datastore.DatastoreOptions;
import com.google.cloud.datastore.Entity;
import com.google.cloud.datastore.EntityQuery;
import com.google.cloud.datastore.Key;
import com.google.cloud.datastore.PathElement;
import com.google.cloud.datastore.Query;
import com.google.cloud.datastore.QueryResults;
import com.google.cloud.datastore.StructuredQuery;
import io.quartic.catalogue.StorageBackend;
import io.quartic.catalogue.api.model.DatasetConfig;
import io.quartic.catalogue.api.model.DatasetCoordinates;

import java.io.IOException;
import java.util.Map;

/**
 * GoogleDatastoreBackend
 *
 * Note that according to https://cloud.google.com/datastore/docs/concepts/structuring_for_strong_consistency
 * strong consistency requires structuring of your data within Google Datastore. In particular only ancestor queries
 * are guaranteed to be strongly consistent. Since we don't want to have to worry about inconsistent state here, we
 * put all of our dataset entities under an ancestor catalogue entity. This should give us strong consistency according
 * to the docs.
 */
public class GoogleDatastoreBackend implements StorageBackend {
    private static final String ANCESTOR = "ancestor";
    private static final String ANCESTOR_KIND = "catalogue";
    private static final String NAMESPACE_KIND = "namespace";
    private static final String DATASET_KIND = "dataset";
    private final Datastore datastore;
    private final Key ancestorKey;
    private final EntitySerDe entitySerDe;

    public GoogleDatastoreBackend(Datastore datastore) {
        this.datastore = datastore;
        this.ancestorKey = datastore.newKeyFactory()
                .setKind(ANCESTOR_KIND)
                .newKey(ANCESTOR);
        this.entitySerDe = new EntitySerDe(this::key);
        datastore.put(Entity.newBuilder().setKey(ancestorKey).build());
    }

    public static GoogleDatastoreBackend remote(String projectId, String namespace) {
        Datastore datastore = DatastoreOptions.newBuilder()
                .setNamespace(namespace)
                .setProjectId(projectId)
                .build()
                .getService();

        return new GoogleDatastoreBackend(datastore);
    }

    @Override
    public DatasetConfig get(DatasetCoordinates coords) throws IOException {
        Entity entity = datastore.get(key(coords));
        return entitySerDe.entityToDataset(entity);
    }

    @Override
    public void put(DatasetCoordinates coords, DatasetConfig config) throws IOException {
        datastore.runInTransaction(readerWriter -> readerWriter.put(entitySerDe.datasetToEntity(coords, config)));
    }

    @Override
    public void remove(DatasetCoordinates coords) throws IOException {
        datastore.delete(key(coords));
    }

    @Override
    public boolean contains(DatasetCoordinates coords) throws IOException {
        return datastore.get(key(coords)) != null;
    }

    @Override
    public Map<DatasetCoordinates, DatasetConfig> getAll() throws IOException {
        EntityQuery query = Query.newEntityQueryBuilder()
                .setKind(DATASET_KIND)
                .setFilter(StructuredQuery.PropertyFilter.hasAncestor(ancestorKey))
                .build();

        Map<DatasetCoordinates, DatasetConfig> datasets = Maps.newHashMap();
        QueryResults<Entity> results = datastore.run(query);
        while (results.hasNext()) {
            Entity entity = results.next();
            datasets.put(
                    new DatasetCoordinates(entity.getKey().getParent().getName(), entity.getKey().getName()),
                    entitySerDe.entityToDataset(entity)
            );
        }
        return datasets;
    }

    @Override
    public HealthCheck.Result healthCheck() {
        if(datastore.get(ancestorKey) != null) {
            return HealthCheck.Result.healthy();
        }
        else {
            return HealthCheck.Result.unhealthy("can't find ancestor key");
        }
    }

    private Key key(DatasetCoordinates coords) {
        return datastore.newKeyFactory()
                .addAncestors(
                        PathElement.of(ANCESTOR_KIND, ANCESTOR),
                        PathElement.of(NAMESPACE_KIND, coords.getNamespace().getNamespace())
                )
                .setKind(DATASET_KIND)
                .newKey(coords.getId().getUid());
    }
}

File: catalogue/src/main/kotlin/io/quartic/catalogue/datastore/GoogleDatastoreBackend.kt
package io.quartic.catalogue.datastore

import com.codahale.metrics.health.HealthCheck.Result
import com.google.cloud.datastore.*
import io.quartic.catalogue.StorageBackend
import io.quartic.catalogue.api.model.DatasetConfig
import io.quartic.catalogue.api.model.DatasetCoordinates
import io.quartic.common.logging.logger

/**
 * GoogleDatastoreBackend

 * Note that according to https://cloud.google.com/datastore/docs/concepts/structuring_for_strong_consistency
 * strong consistency requires structuring of your data within Google Datastore. In particular only ancestor queries
 * are guaranteed to be strongly consistent. Since we don't want to have to worry about inconsistent state here, we
 * put all of our dataset entities under an ancestor catalogue entity. This should give us strong consistency according
 * to the docs.
 */
class GoogleDatastoreBackend(private val datastore: Datastore) : StorageBackend {
    private val LOG by logger()
    private val ancestorKey = datastore.newKeyFactory().setKind(ANCESTOR_KIND).newKey(ANCESTOR)
    private val entitySerDe = EntitySerDe { this.key(it) }

    init {
        datastore.put(Entity.newBuilder().setKey(ancestorKey).build())
    }

    override fun get(coords: DatasetCoordinates) = entitySerDe.entityToDataset(datastore.get(key(coords)))

    override fun set(coords: DatasetCoordinates, config: DatasetConfig) {
        datastore.runInTransaction { readerWriter -> readerWriter.put(entitySerDe.datasetToEntity(coords, config)) }
    }

    override fun remove(coords: DatasetCoordinates) {
        datastore.delete(key(coords))
    }

    override fun contains(coords: DatasetCoordinates) = datastore.get(key(coords)) != null

    override fun getAll(): Map<DatasetCoordinates, DatasetConfig> {
        val query = Query.newEntityQueryBuilder()
                .setKind(DATASET_KIND)
                .setFilter(StructuredQuery.PropertyFilter.hasAncestor(ancestorKey))
                .build()

        val datasets = mutableMapOf<DatasetCoordinates, DatasetConfig>()
        val results = datastore.run(query)
        while (results.hasNext()) {
            val entity = results.next()

            if (entity.key.ancestors.size == 1) {
                LOG.info("Skipping un-namespaced dataset '" + entity.key.name + "'")
            } else {
                datasets.put(
                        DatasetCoordinates(entity.key.parent.name, entity.key.name),
                        entitySerDe.entityToDataset(entity)
                )
            }
        }
        return datasets
    }

    override fun healthCheck() = if (datastore.get(ancestorKey) != null) {
        Result.healthy()
    } else {
        Result.unhealthy("can't find ancestor key")
    }

    private fun key(coords: DatasetCoordinates) = datastore.newKeyFactory()
            .addAncestors(
                    PathElement.of(ANCESTOR_KIND, ANCESTOR),
                    PathElement.of(NAMESPACE_KIND, coords.namespace.namespace)
            )
            .setKind(DATASET_KIND)
            .newKey(coords.id.uid)

    companion object {
        private val ANCESTOR = "ancestor"
        private val ANCESTOR_KIND = "catalogue"
        private val NAMESPACE_KIND = "namespace"
        private val DATASET_KIND = "dataset"

        fun remote(projectId: String, namespace: String) = GoogleDatastoreBackend(DatastoreOptions.newBuilder()
                .setNamespace(namespace)
                .setProjectId(projectId)
                .build()
                .service
        )
    }
}
